mypy/types.py

```diff
@@ -858,17 +858,18 @@ def make_simplified_union(items: List[Type], line: int = -1, column: int = -1) -
                     all_items.append(typ)
             items = all_items
 
-        if any(isinstance(typ, AnyType) for typ in items):
-            return AnyType()
-
         from mypy.subtypes import is_subtype
         removed = set()  # type: Set[int]
         for i, ti in enumerate(items):
             if i in removed: continue
             # Keep track of the truishness info for deleted subtypes which can be relevant
             cbt = cbf = False
             for j, tj in enumerate(items):
-                if i != j and is_subtype(tj, ti):
+                # Attempt to only combine true subtypes by avoiding types containing Any.
+                # TODO: Properly exclude generics and functions.
+                tj_is_anylike = (isinstance(tj, AnyType) or
```

Comment:
What if there are multiple `Any` types in the union to simplify? I think that we should remove duplicates away at least, perhaps by using `is_any_type`. So `Union[Any, Any, int]` would be simplified into `Union[Any, int]`.